src/librustc_data_structures/obligation_forest/mod.rs

```diff
@@ -30,28 +32,64 @@ pub struct ObligationForest<O> {
     /// incomplete children). During processing, some of those nodes
     /// may be changed to the error state, or we may find that they
     /// are completed (That is, `num_incomplete_children` drops to 0).
-    /// At the end of processing, those nodes will be removed by a
-    /// call to `compress`.
+    /// At the end of processing, those nodes will be removed (or
+    /// marked as removed if used in earlier snapshots) by a call to
+    /// `compress`.
     ///
     /// At all times we maintain the invariant that every node appears
     /// at a higher index than its parent. This is needed by the
     /// backtrace iterator (which uses `split_at`).
     nodes: Vec<Node<O>>,
-    snapshots: Vec<usize>
+    snapshots: Vec<Snapshot>
 }
 
-pub struct Snapshot {
-    len: usize,
-}
+// We could implement Copy here, but that tastes weird.
```

Comment:
Nit: This comment is funny, but you could also say "but we expect each snapshot to be consumed exactly once".